examples/ecdsa.c

```diff
@@ -0,0 +1,121 @@
+/*************************************************************************
+ * Copyright (c) 2020-2021 Elichai Turkel                                *
+ * Distributed under the CC0 software license, see the accompanying file *
+ * EXAMPLES_COPYING or https://creativecommons.org/publicdomain/zero/1.0 *
+ *************************************************************************/
+
+#include <stdio.h>
+#include <assert.h>
+#include <string.h>
+
+#include "random.h"
+#include "secp256k1.h"
+
+
+
+int main(void) {
+    unsigned char msg_hash[32] = {0}; /* This must be a hash of the message. otherwise ECDSA is easily broken. */
+    unsigned char seckey[32];
+    unsigned char randomize[32];
+    unsigned char compressed_pubkey[33];
+    unsigned char serialized_signature[64];
+    size_t len;
+    int is_signature_valid;
+    secp256k1_pubkey pubkey;
+    secp256k1_ecdsa_signature sig;
+    /* The docs in secp256k1.h above the `secp256k1_ec_pubkey_create` function
+     * say: "pointer to a context object, initialized for signing" And the docs
+     * above the `secp256k1_ecdsa_verify` function say: "a secp256k1 context
+     * object, initialized for verification" which is why we create a context
+     * for both signing and verification with the SECP256K1_CONTEXT_SIGN and
+     * SECP256K1_CONTEXT_VERIFY flags. */
+    secp256k1_context* ctx = secp256k1_context_create(SECP256K1_CONTEXT_SIGN | SECP256K1_CONTEXT_VERIFY);
```

Comment:
nit: Could assert that the context was created successfully (here and in the two other examples)? I.e. `assert(ctx != NULL);`
IIUC returning NULL can only happen in the unlikely case that allocating dynamic memory fails, but it can't hurt. We also do that check in Bitcoin Core: https://github.com/bitcoin/bitcoin/blob/8f137e69caeb2a2ffe1aa930bd6fbc49cee4087c/src/key.cpp#L394-L395

Reply:
If that was the case then I would agree that the examples should do the check. However, `context_create` calls `checked_malloc` which will call the "error callback" if the result of malloc is NULL. The default behavior of the error callback is crashing. It could theoretically be overriden by the user to not crash (`context_set_error_callback`) but that would be inadvisable given that the API docs state that after the callback the originally called API function can do anything, in particular its return values are undefined. That's why the API doc for `secp256k1_context_create` says that it returns a "context object" and not a "context object or NULL".